src/System.Text.Json/src/System/Text/Json/JsonDocument.cs

```diff
@@ -0,0 +1,436 @@
+                    {
+                        database.SetHasComplexChildren(parentLocation);
+                    }
+
+                    parentLocation = database.Length;
+
+                    if (inArray)
+                    {
+                        arrayItemsCount++;
+                    }
+
+                    numberOfRowsForValues++;
+                    database.Append(tokenType, tokenStart, DbRow.UnknownSize, inObject);
+                    var row = new StackRow(numberOfRowsForMembers + 1);
+                    stack.Push(row);
+                    numberOfRowsForMembers = 0;
+                }
+                else if (tokenType == JsonTokenType.EndObject)
+                {
+                    parentLocation = -1;
+
+                    int rowIndex = database.FindIndexOfFirstUnsetSizeOrLength(JsonTokenType.StartObject);
+
+                    numberOfRowsForValues++;
+                    numberOfRowsForMembers++;
+                    database.SetLength(rowIndex, numberOfRowsForMembers);
+
+                    database.Append(tokenType, tokenStart, reader.ValueSpan.Length, false);
+                    database.SetNumberOfRows(rowIndex, numberOfRowsForMembers);
+
+                    StackRow row = stack.Pop();
+                    numberOfRowsForMembers += row.SizeOrLength;
+                }
+                else if (tokenType == JsonTokenType.StartArray)
+                {
+                    if (parentLocation != -1)
+                    {
+                        database.SetHasComplexChildren(parentLocation);
+                    }
+
+                    parentLocation = database.Length;
+
+                    if (inArray)
+                    {
+                        arrayItemsCount++;
+                    }
+
+                    numberOfRowsForMembers++;
+                    database.Append(tokenType, tokenStart, DbRow.UnknownSize, inObject);
+                    var row = new StackRow(arrayItemsCount, numberOfRowsForValues + 1);
+                    stack.Push(row);
+                    arrayItemsCount = 0;
+                    numberOfRowsForValues = 0;
+                }
+                else if (tokenType == JsonTokenType.EndArray)
+                {
+                    parentLocation = -1;
+
+                    int rowIndex = database.FindIndexOfFirstUnsetSizeOrLength(JsonTokenType.StartArray);
+
+                    numberOfRowsForValues++;
+                    numberOfRowsForMembers++;
+                    database.SetLength(rowIndex, arrayItemsCount);
+                    database.SetNumberOfRows(rowIndex, numberOfRowsForValues);
+
+                    database.Append(tokenType, tokenStart, reader.ValueSpan.Length, false);
+                    StackRow row = stack.Pop();
+                    arrayItemsCount = row.SizeOrLength;
+                    numberOfRowsForValues += row.NumberOfRows;
+                }
+                else if (tokenType == JsonTokenType.PropertyName)
```

Comment:
Are these ordered from most common to least common?  Since the further you get here the more checks you need to do as part of parsing, it might be interesting to gather some stats on frequency on common documents and order accordingly.  Or maybe it doesn't matter.